### PR TITLE
Add root cause analysis API with multidimension as a kind

### LINKED ISSUE #16984: [Cognitive Services - Anomaly Detector] Multidimensional Root Cause Analysis
• Key contact for the review:  suxi@microsoft.com
• Whether this is a new or existing API:  This is a set of new API
• Whether or not your service already been deployed:   Hasn't been deployed
• A brief description of the material to be reviewed/change:  This new feature from Anomaly Detector service is Multidimensional Root Cause Analysis (MDRCA). It can discover the root causes of problems to identify appropriate solutions within your complex metric systems such as ROI, CTR, through detecting the abnormal dimension(s) or abnormal dimension combinations when there is a big spike or drop in your metrics system, without knowing any machine learning knowledge, labeled data, or even anomaly information.
• A link to the PR/issue in  azure/azure-rest-api-specs [link](https://github.com/Azure/azure-rest-api-specs/pull/18037)

### PR DESCRIPTION
**Pull request by suxi-ms**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json [hunk 1]
```diff
@@ -0,0 +1,2162 @@
...
+    "version": "2.0-preview"
+  },
+  "paths": {
+    "/rootCauseAnalysis/dataSets": {
+      "put": {
```
Comment (suxi-ms) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:10]: We don't support update for the resource as change of schema of data set makes existing data unmatched.  From the restful API guide, we find POST is also supported when URL of created resource is returned. We plan to use the display name to make it idemdepotent.  Will this work?

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json [hunk 2]
```diff
@@ -0,0 +1,2162 @@
...
+        },
+        "displayName": {
+          "description": "Root cause analysis data set name",
+          "type": "string"
+        }
+      },
+      "discriminator": "kind",
+      "example": {
+  "granularity": "daily",
+  "customInterval": 0,
```
Comment (suxi-ms) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:934]: We use this customInterval to align with existing MVAD API set. It describe the unit of granularity. For example, when graularity is daily, customInterval is 5, it means  5 days granularity.

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+      "in": "header"
+    }
+  },
+  "security": [
+    {
+      "apiKeyHeader": []
+    }
+  ],
+  "x-ms-parameterized-host": {
+    "hostTemplate": "{Endpoint}/anomalydetector/v2.0-preview",
```
Comment (bexxx) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json:21]: >v2.0-preview [](http://example.com/codeflow?start=48&length=12)

There was a push to move the api version to a query parameter for all new major versions on CS.

Not sure how strict this is now handled in CS.

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+          },
+          {
+            "value": "daily"
+          },
+          {
+            "value": "hourly"
+          },
+          {
+            "name": "perMinute",
+            "value": "minutely"
```
Comment (bexxx) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json:214]: >minutely [](http://example.com/codeflow?start=22&length=8)

I am not a native speaker, but this seems not to mean every minute.

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+          },
+          {
+            "name": "perSecond",
+            "value": "secondly"
```
Comment (bexxx) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json:218]: >secondly [](http://example.com/codeflow?start=22&length=8)

also this does most probably not mean every second

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json [hunk 6]
```diff
@@ -0,0 +1,2143 @@
...
+        },
+        "x-ms-examples": {
+          "Delete a root cause analysis data set": {
+            "$ref": "./examples/DeleteRootCauseAnalysisADataSet.json"
+          }
+        }
+      }
+    },
+    "/rootCauseAnalysis/dataSets/{dataSetId}:upload": {
+      "post": {
```
Comment (weidongxu-microsoft) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:212]: What is the options, if user failed to receive the response of the POST (due to timeout etc.)? Will retry work?

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+  },
+  "securityDefinitions": {
+    "apiKeyHeader": {
+      "type": "apiKey",
+      "name": "Ocp-Apim-Subscription-Key",
+      "in": "header"
```
Comment (mikekistler) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json:12]: Please include a description in the securityDefinition that can be shown in the generated REST API reference docs, e.g
```
 "description": "The secret key for your Azure Cognitive Services subscription."
```

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json [hunk 8]
```diff
@@ -0,0 +1,2143 @@
...
+          {
+            "in": "query",
+            "name": "maxpagesize",
+            "description": "The maximum number of items in one page",
+            "type": "integer",
+            "format": "int32"
+          },
+          {
+            "in": "query",
+            "name": "`filter`",
```
Comment (mikekistler) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:105]: Please don't use special characters in parameter names.  This should be just "filter".

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json [hunk 9]
```diff
@@ -0,0 +1,2143 @@
...
+          ],
+          "type": "string",
+          "x-ms-enum": {
+            "name": "RootCauseAnalysisType",
+            "modelAsString": true
+          }
+        }
+      },
+      "discriminator": "kind"
+    },
```
Comment (mikekistler) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:1411]: Here is another schema with a discriminator with only one "kind". I really want to understand why/if this is needed.

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json [hunk 10]
```diff
@@ -0,0 +1,2143 @@
...
+      },
+      "x-ms-discriminator-value": "multidimensional"
+    },
+    "RootCauseAnalysisDataMetaListResponse": {
+      "required": [
+        "value"
+      ],
+      "type": "object",
+      "properties": {
+        "@nextLink": {
```
Comment (mikekistler) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:1454]: All property names should be camel-case with no special characters.  Please use "nextLink".

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json [hunk 11]
```diff
@@ -0,0 +1,2143 @@
...
+        }
+      },
+      "x-ms-discriminator-value": "multidimensional"
+    }
+  },
+  "securityDefinitions": {
+    "apiKeyHeader": {
+      "type": "apiKey",
+      "name": "Ocp-Apim-Subscription-Key",
+      "in": "header"
```
Comment (mikekistler) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:2115]: Please include a description in the securityDefinition that can be shown in the generated REST API reference docs.
```
 "description": "The secret key for your Azure Cognitive Services subscription."
```

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json [hunk 12]
```diff
@@ -0,0 +1,2143 @@
...
+            "$ref": "./examples/CreateRootCauseAnalysisDataSet.json"
+          }
+        }
+      },
+      "get": {
+        "tags": [
+          "RootCauseAnalysisDataSet"
+        ],
+        "summary": "List root cause analysis data set",
+        "operationId": "ListRootCauseAnalysisDataSet",
```
Comment (mikekistler) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:77]: I think this is listing all the data sets, right?
```suggestion
        "summary": "List root cause analysis data sets",
        "operationId": "ListRootCauseAnalysisDataSets",
```

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json [hunk 13]
```diff
@@ -0,0 +1,2143 @@
...
+          "application/json"
+        ],
+        "parameters": [
+          {
+            "in": "path",
+            "name": "dataSetId",
+            "description": "The root cause analysis data set unique id",
+            "required": true,
+            "type": "string",
+            "format": "uuid"
```
Comment (mikekistler) [inline specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:154]: You should specify the maximum length (`maxLength`) and allowed characters (`pattern`) for path parameters. Path parameters should typically be limited to alphabetic, numeric, and a few special characters like "-" and "_".

**Review by mikekistler — CHANGES_REQUESTED** [review]
I noted a number of issues I think should be addressed.

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json (hunk 9 above)
Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:1411]: kind is a field to describe what kind of RCA this instance is . Currently, the kind is multidimensional root cause analysis. In the future, we plan to provide cross metric root cause analysis APIs

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json (hunk 10 above)
Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:1454]: yes, updated in new PR

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json (hunk 8 above)
Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:105]: updated in new PR

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json (hunk 3 above)
Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json:21]: updated version in new PR

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json (hunk 4 above)
Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json:214]: This is from the existing API. If it is not a good option, maybe we need to discuss and fix this problem in a new thread.

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json (hunk 5 above)
Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/AnomalyDetector.json:218]: same as above

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json (hunk 6 above)
Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:212]: We have validation in API that same partition Id cannot be uploaded more than once

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json [hunk 14]
```diff
@@ -0,0 +1,2143 @@
...
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Success",
+            "schema": {
+              "$ref": "#/definitions/RootCauseAnalysisDataSetList"
+            }
+          },
+          "400": {
```
[the remark this answers is not in the document]
Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:117]: will use default

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json [hunk 15]
```diff
@@ -0,0 +1,2143 @@
...
+              "$ref": "#/definitions/MultiDimRCAListDataSetErrorResponse"
+            },
+            "headers": {
+              "x-ms-error-code": {
+                "description": "String identical to error code.",
+                "type": "string"
+              }
+            }
+          }
+        },
```
[the remark this answers is not in the document]
Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:129]: updated in new PR

File: specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json (hunk 12 above)
Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/AnomalyDetector/preview/v2.0-preview/RootCauseAnalysis.json:77]: updated in new PR

**Pull request closed without merging**